#01.Import Libraries and Data

In [5]:
#import libraries
import pandas as pd
import numpy as np
import os

In [6]:
#import products.csv
#set path
path = r'/Users/juliewhitton/Desktop/Instacart Basket Analysis'

In [7]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 
                                   'products.csv'), index_col = False)

In [8]:
#import wrangled orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 
                                   'orders_wrangled.csv'), index_col = False)

#02.Mixed Type Data

#03.Missing Values

In [9]:
#check df_prods for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
#view missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
#check rows of df_prods
df_prods.shape

(49693, 5)

In [13]:
#create new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull()
                          == False]

In [14]:
#check rows of df_prods_clean
df_prods_clean.shape

(49677, 5)

#04.Duplicates

In [15]:
#check for dupes
df_dupes = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
#call df_dupes
df_dupes

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
#check # of rows
df_prods_clean.shape

(49677, 5)

In [18]:
#create new data frame sans dupes
df_prods_clean_no_dupes = df_prods_clean.drop_duplicates()

In [19]:
#check # of rows
df_prods_clean_no_dupes.shape

(49672, 5)

In [20]:
#export dupe-free df
df_prods_clean_no_dupes.to_csv(os.path.join(path,
                                          '02 Data', 'Prepared Data',
                                          'prods_checked.csv'))

#05.Task 4.5

In [22]:
#check df_ords
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   time_of_order  days_since_prior_order  
0              8                     NaN  
1              7                    15.0  
2             12                    21.0  
3              7                    29.0  
4             15                    28.0

In [26]:
#check df_ords with describe()
df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       time_of_order  days_since_prior_order  
count   3.421083e+06            3.214874e+06  
mean    1.345202e+01            1.111484e+01  
std     4.226088e+00            9.206737e+00  
min     0.000000e+00            0.000000e+00  
25%     1.000000e+01            4.000000e+00  
50%     1.300000e+01            7.000000e+00  
75%     1.600000e+01            1.500000e+01  
max     2.300000e+01            3.000000e+01

#06.Mixed Type Data df_ords

In [27]:
#check for mixed type data in df_ords
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) !=
            df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ords[weird]) > 0:
        print(col)

#07.Missing Values in df_ords

In [30]:
#check for missing values in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
time_of_order                  0
days_since_prior_order    206209
dtype: int64

In a markdown cell, report findings and possible explanation:
There are 206,209 values missing in the 'days_since_prior_order' column. This could be due to single-use customers.

In [31]:
#address the missing values
#create new df subset
#check # of rows
df_ords.shape

(3421083, 6)

Explain why you chose this method for missing values:
Even though there were more than 200k values missing, it is not a large enough percentage of fields to warrant deleting the column. 

#08.Duplicates df_ords

In [32]:
#check df_ords for dupes
df_ords_dupes = df_ords[df_ords.duplicated()]


In [33]:
#view dupes
df_ords_dupes

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, time_of_order, days_since_prior_order]
Index: []

In [34]:
#create a new df w/out dupes
df_ords_checked = df_ords

In [35]:
#check # of rows
df_ords_checked.shape

(3421083, 6)

Explain your method of excluding duplicates:
I attempted to create a new data frame using df.drop_dupicates() but when I checked the number of rows, it is identical to the number I got before attempting to delete the duplicate row. 

In [36]:
#export checked data
#df_prods was exported before the beginning of this task
df_ords_checked.to_csv(os.path.join(path,
                                    '02 Data', 'Prepared Data',
                                    'ords_checked.csv'))

In [37]:
df_ords_checked.shape

(3421083, 6)